### PR TITLE
Do not refresh the names of implicit arguments.

### LINKED ISSUE #6785: The initial implicit argument names can differ from the specified names
#### Version

8.8 master hash 8dd6d09 and 8.7.0

#### Description of the problem

```
Definition foo (a a : nat) : nat := 3.
Fail Arguments foo a a.
(*
To rename arguments the "rename" flag must be specified.
Argument a0 renamed to a.
*)

Fail Arguments foo a a : rename.
(* Some argument names are duplicated: a *)

Arguments foo a a0 : assert.
Arguments foo a b : rename.
```

This behavior seems confusing and fragile. Would it be possible to allow repeated names for argument specifications, and use the first one on an implicit application `(a := 2)`? As opposed to actual binders, I don't see a problem with shadowing uses in later types. One issue I do see is that queries like `About` and `Print` will show the renamed varaibles, but that is the same situation as after an explicit `Arguments foo c d`, for which `About` gives
```
foo : nat -> forall a0 : nat, a0 = a0
Arguments are renamed to c, d
...
```
Having instead ```Arguments are named a, a``` always printed doesn't seem too different.

---

One effect of the option in #6582 is to also mangle default argument names, since they get refreshed (with respect to the local env. not the global env.). This caused problems working on #6781.
I think that changing the above behavior would be an improvement to the system, making it easier to understand and work with, but the argument could also be made that #6582 is too agressive here.

### LINKED ISSUE #12001: Name mangling mangles explicitly given argument names
#### Description of the problem

Name mangling via `-mangle-names` seems to mangle each and every argument name, even those that are explicitly given:
```coq
Lemma leibniz_equiv_iff {A : Type} (x y : A) : True.
Proof. tauto. Qed.
About leibniz_equiv_iff.
```
prints
```
leibniz_equiv_iff : forall {mangled_0 : Type}, mangled_0 -> mangled_0 -> True
```

That currently makes it rather useless as it breaks uses like `leibniz_equiv_iff (A:=...)`.

#### Coq Version

```
$ coqc --version
The Coq Proof Assistant, version 8.12+alpha (April 2020)
compiled on Apr 2 2020 8:51:47 with OCaml 4.08.1
```
But that's not very useful for dev builds, so let's ask opam:
```
<><> Version-specific details <><><><><><><><><><><><><><><><><><><><><><><><><>
version      dev
repository   coq-core-dev-test
source-hash  971c8e12
```

### PR DESCRIPTION
**Pull request by jashug**
Try just going with the user-given names, and not worrying about what happens with repeated names or anonymous implicits.
(Support for anonymous implicits is due to herbelin in coq/coq#11098.)

This avoids unnecessarily refreshing the default name to `name0` in several cases in the CI.

This is another step towards checking the standard library with `-mangle-names`.
Closes coq/coq#6785
Closes coq/coq#12001
Unblocks coq/stdlib#50

This PR should not change behaviour in the absence of repeated names.
Since repeated names are already a poorly handled corner case, I would recommend changing binder names to avoid overlap in the case of a change in behavior.

The most common case caught by the CI was that a name like `eq0` now becomes `eq`, and that causes `Arguments` to complain about renaming. The fix is simply to add the `rename` flag or change to use the un-suffixed name.

Since anonymous implicits and implicits with repeated names can already happen, I think this is unlikely to cause too many new problems, though it might exacerbate existing ones. However, I already had to fix one newly possible anomaly, so I can't be too confident.

**Kind:** bug fix?
I'm not really sure. It closes one issue by someone other than myself, so probably reasonable to call it a bug fix.

- [x] Added / updated test-suite
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

Overlays:
- [x] [color](https://github.com/fblanqui/color/pull/29)
- [x] [math-comp](https://github.com/math-comp/math-comp/pull/550)
- [x] [odd-order](https://github.com/math-comp/odd-order/pull/25)

### CODE REVIEW

File: test-suite/output/Implicit.out
Comment (herbelin) [inline test-suite/output/Implicit.out:8]: Would it make sense here to go one step further and to consider that the second argument `x` of `d2` is unavailable by name (forcing a `_` in `Arguments`), on the basis that the user deliberatedly chose `x` as first argument of `d2`?

Or to warn at the time of defining `d2`?

File: vernac/comArguments.ml
Comment (herbelin) [inline vernac/comArguments.ml:193]: Not a big deal, but why the move?

**Review by herbelin — APPROVED** [review]
It is good news that the incompatibilities are few. I'm supporting this PR which will help me also for #11272 (see also #11510).

I may be the assignee, but, being related to #12001, @maximedenes may be a better assignee if ever he's willing.

File: test-suite/output/Implicit.out
Comment (jashug) [reply test-suite/output/Implicit.out:8]: Something like that does indeed sound good, yes.

The code change required is more than just doing the deduplication in `compute_implicits_names_gen`; anonymous implicit arguments there adopt the name from the manual implicit information somewhere in `prepare_implicits` (I think), and I haven't put in the time to re-understand the logic of `prepare_implicits` in the light of anonymous implicits.

Warning might be tricky, since for example a `Proper (R1 ==> R2 ==> R3)` instance gets unfolded to find argument names of `x y _ x y _`, which might warn without special handling to silence warnings after unfolding.

I might also advocate for keeping only the *last* argument of the same name, since that notion is stable under discharging sections, but that is unintuitive in it's own way on the `Proper` example.

File: vernac/comArguments.ml
Comment (jashug) [reply vernac/comArguments.ml:193]: After the move, `Arguments` only checks for duplicates in the explicitly user-given names, not in the final list of names.
For example, without this change:
```coq
Definition foo (a a : nat) := a.
Arguments foo _ _ : assert.
```
fails with `Some argument names are dupliated: a`, and this happened several places in the CI.

I'm not entirely happy with this change, as it probably lets through some unexpected corner cases, but it is a minimal change that reduces incompatibility with existing developments, and doesn't completely get rid of the (useful) error.
The `Arguments` command is probably due to be looked at closely sometime soon, but maybe after more of your changes to implicits go in.

**Comment by jashug** [comment]
Overlay PRs submitted, not yet merged. See initial post for links (now updated from pointing to the diffs to pointing to the PRs).

**Comment by jashug** [comment]
@maximedenes I would like to hear any feedback you might have on this change before pinging math-comp to merge overlays.
Note that this would close #12001.

**Comment by maximedenes** [comment]
@jashug Sorry, this PR went below my radar. Having a look now.

File: test-suite/output/Implicit.out
Comment (maximedenes) [reply test-suite/output/Implicit.out:8]: I think it's ok to tolerate duplicated names, like OCaml does for labels.

File: vernac/comArguments.ml
Comment (maximedenes) [reply vernac/comArguments.ml:193]: Since it is now allowed internally, why not let the user repeat names in `Arguments`?

**Review by maximedenes — COMMENTED** [review]
This PR seems like a good step to unblock the problem. However, since repeated names are now allowed, why not remove the `Arguments` check? It seems to me that the only difference it makes is that it now takes two calls to `Arguments` to define repeated names instead of one, which looks like an arbitrary restriction.

File: vernac/comArguments.ml
Comment (jashug) [reply vernac/comArguments.ml:193]: My reasoning was that since there is still no way to *use* repeated names in `(name := arg)` applications, users almost never want duplicated names and so the error is useful. However, taking another look, that seems overly paternalistic; names can also simply be documentation, and we allow naming non-implicit arguments. Since the check for duplicates is already incomplete, it is probably better to be more consistent and remove this check.

**Review by maximedenes — APPROVED** [review]

**Comment by maximedenes** [comment]
@jashug The PR looks good and ready (modulo CI, of course). I think you can go ahead and ping people on overlay PRs (if they are all backward compatible). Thanks a lot for this patch!

**Pull request merged**
